### DISCUSSION TITLE
Macros

### DISCUSSION THREAD

**Discussion by IHateYourCode** [texstudio-org/texstudio#1841] [Ideas]
I'm considering going through all macros used in
**TeXStudio** and adjusting them to all use the same style.

Please let me know if you take issue
with the following style I'm proposing:

---

#### Rewrite **header guards** with:

```c++
#ifndef Header_Editor_Factory
#define Header_Editor_Factory
```
This is a more readable version compared to the
different styles used throughout texstudio, such as:
`adwaitascollbarengine_h` or `ABOUTDIALOG_H`

---

#### 'Local' macros vs 'Global' macros

As macros are by default 'global', overlaps can lead to problems.
To separate the two 'types' I would like to write all 'global' ones as
`Capitalized_Names` and local ones `completely_lowercase`.

On that note, I would like to ensure that all local 
macros are definitely `#undef` afterwards.

---

#### Safe Macros

I would also like to ensure that macros use proper
argument wrapping to minimize error potentials.

```c++
#define macro(value) value
```
becomes
```c++
#define macro(value) (value)
```

---

#### Constexpr > Macros

I would like to replace macros with constexpr where
appropriate to again minimize error potential.

**Comment by sunderme** [comment]
sounds fine

**Reply by IHateYourCode** [reply]
[Finished]: https://github.com/IHateYourCode/texstudio/commit/3712326bf57233d1cc117556b44d5c5b3850dfa0
[Branch]: https://github.com/IHateYourCode/texstudio/tree/Macro-Refactor
Magnificent
Actually already [Finished] the easy (though tedious) part.
For future reference, here's the [Branch].

**Comment by sunderme** [comment]
Looking at the code as it presents itself, I disagree with setting macros all lowercase.
It should be easily distinguishable to normal functions.
Google code style propagate: 
* don't use macros
* if you do, all uppercase.

This has been the norm for nearly all cpp projects I have seen.
So, please don't change macros that already follow that idea.
Local macros,where definition and use is close to each other, may be tolerated.
For the others, I may revert the changes, yet undecided.
In general inline functions are preferable, so changes to that are okay (with lowercase function names)
And so are const.

**Comment by sunderme** [comment]
A second aspect: don't touch imported 3rd party libs
We don't test them, they maybe overwritten without notice and we do not want to get into them too much.
Those libraries are:
* diff_match*
* adwaita
* hunspell
* quazip
* xkb
* synctex

They have already been touched, and I hope no error was introduced, let's keep fingers crossed.

**Comment by sunderme** [comment]
Again, after this excercise today, there are really more beneficial things to do in the code than renaming macros.
Another field is fixing compiler/clazy warnings.
If I have to invest hours of my time to avoid breaking code (and it would have broken at several places) with little benefits (removing macros I consider a benefit, but the renaming is not), I would rather spend it on something which improves txs in the long run.
